src/rust/engine/fs/fs_util/src/main.rs

```diff
@@ -115,12 +144,52 @@ fn execute(top_match: clap::ArgMatches) -> Result<(), String> {
           let posix_fs = Arc::new(make_posix_fs(args.value_of("source").unwrap()));
           let (fingerprint, size_bytes) =
             save_directory(store, posix_fs, Arc::new(fs::Dir(PathBuf::from("."))))
-              .wait()?;
+              .wait()
+              .map_err(|e| (e, 1))?;
           Ok(println!("{} {}", fingerprint, size_bytes))
         }
+        ("cat-proto", Some(args)) => {
+          let fingerprint = Fingerprint::from_hex_string(args.value_of("fingerprint").unwrap())
+            .map_err(|e| (e, 1))?;
+          match store.load_directory_proto_bytes(&fingerprint).map_err(
+            |e| (e, 1),
+          )? {
+            Some(bytes) => {
+              io::stdout().write(&bytes).unwrap();
+              Ok(())
+            }
+            None => Err((
+              format!(
+                "Directory with fingerprint {} not found",
+                fingerprint
+              ),
+              2,
+            )),
+          }
+        }
         (_, _) => unimplemented!(),
       }
     }
+    ("cat", Some(args)) => {
+      let fingerprint = Fingerprint::from_hex_string(args.value_of("fingerprint").unwrap())
+        .map_err(|e| (e, 1))?;
+      let v = match store.load_file_bytes(&fingerprint).map_err(|e| (e, 1))? {
```

Comment:
This block makes me kind of sad... I was hoping to be able to write something like:
```
let bytes: Option<Vec<u8>> = store.load_file_bytes(&fingerprint)
  .or_else(|| store.load_directory_proto_bytes(&fingerprint)?)
  .map_err(|e| (e, 1))?
```

But I couldn't work out a way to make the Results propagate properly

Reply:
Since both of the methods have the same signature, something like the following might work:
```
store.load_file_bytes(&fingerprint)
  .and_then(|res| {
    res.map(|v| Ok(v)).or_else(store.load_directory_proto_bytes(&fingerprint)))
  })
  .map_err(|e| (e, 1))?;
```
The challenge is definitely with the nested `Result`/`Option`.

Reply:
`map_or_else` appears to be the way to do this... I've pushed a second commit which uses that, but I'm not sure it's more readable... Thoughts?